codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/protocol/aws/SerializeMiddleware.java

```diff
@@ -0,0 +1,223 @@
+                """, Map.of(
+                    "opName", this.serialName,
+                    "handleSerialize", generateHandleSerialize()
+        ));
+    }
+
+    private GoWriter.Writable generateHandleSerialize() {
+        return goTemplate(
+                """
+                        
+                        func (op *$opName:L) HandleSerialize (ctx $context:T, in $input:T, next $handler:T) (
+                        out $output:T, metadata $metadata:T, err error) {
+                        
+                            $body:W
+                        
+                            return next.HandleSerialize(ctx, in)
+                        }
+                        
+                        """, Map.of(
+                        "context", SmithyGoDependency.CONTEXT.interfaceSymbol("Context"),
+                        "input", SmithyGoDependency.SMITHY_MIDDLEWARE.struct("SerializeInput"),
+                        "handler", SmithyGoDependency.SMITHY_MIDDLEWARE.struct("SerializeHandler"),
+                        "output", SmithyGoDependency.SMITHY_MIDDLEWARE.struct("SerializeOutput"),
+                        "metadata", SmithyGoDependency.SMITHY_MIDDLEWARE.struct("Metadata"),
+                        "opName", this.serialName,
+                        "body", generateHandleSerializeBody())
+        );
+    }
+
+    private GoWriter.Writable generateHandleSerializeBody() {
+        return goTemplate("""
+                    $route:W
+    
+                    $serialize:W
+    
+                """,
+                MapUtils.of(
+                        "input", ctx.getSymbolProvider().toSymbol(input),
+                        "request", generator.getApplicationProtocol().getRequestType(),
+                        "route", handleHttp(),
+                        "serialize", handlePayload(),
+                        "errorf", GoStdlibTypes.Fmt.Errorf
+                ));
+    }
+
+    private GoWriter.Writable handleHttp() {
+        return goTemplate("""
+                    input, ok := in.Parameters.($input:P)
+                    if !ok {
+                        return out, metadata, $errorf:T("unexpected input type %T", in.Parameters)
+                    }
+                    _ = input
+                
+                    req, ok := in.Request.($request:P)
+                    if !ok {
+                        return out, metadata, $errorf:T("unexpected transport type %T", in.Request)
+                    }
+                
+                    req.Method = $methodPost:T
+                    req.URL.Path = "/service/$service:L/operation/$operation:L"
+                    req.Header.Set("smithy-protocol", $protocol:S)
+                
+                    $contentTypeHeader:W
+                    $acceptHeader:W
+                
+                """,
+        MapUtils.of(
+                "input", ctx.getSymbolProvider().toSymbol(input),
+                "request", generator.getApplicationProtocol().getRequestType(),
```

Comment:
We know this is a smithyhttp.Request since it's awsjson10 - so you can just embed that symbol here. That will also enable you to get rid of the protocol generator being passed into the middleware.